Reimplement SignatureBinder to support repeated literal variable

haozhun opened the pull request.

cberner commented:
There are some checkstyle failures

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java:
```diff
@@ -217,8 +218,9 @@ public static long fromISO8601Date(ConnectorSession session, @SqlType(StandardTy
+    @LiteralParameters("x")
     @SqlType(StandardTypes.TIME_WITH_TIME_ZONE)
-    public static long timeAtTimeZone(@SqlType(StandardTypes.TIME_WITH_TIME_ZONE) long timeWithTimeZone, @SqlType(StandardTypes.VARCHAR) Slice zoneId)
+    public static long timeAtTimeZone(@SqlType(StandardTypes.TIME_WITH_TIME_ZONE) long timeWithTimeZone, @SqlType("x") Slice zoneId)
```
This isn't right. In fact this should have failed to load. Can you add some validation logic to the signature parsing code path to make sure you can't have a bare literal parameter?

haozhun replied:
done. Added a new commit to parsing validation.

cberner commented on presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java:
```diff
@@ -71,32 +72,32 @@
+    public static final SqlScalarFunction DOUBLE_TO_DECIMAL_CAST = castFunctionToDecimalFrom(DOUBLE.getTypeSignature(), "doubleToShortDecimal", "doubleToLongDecimal");
+    public static final SqlScalarFunction DECIMAL_TO_FLOAT_CAST = castFunctionFromDecimalTo(FLOAT.getTypeSignature(), "shortDecimalToFloat", "longDecimalToFloat");
+    public static final SqlScalarFunction FLOAT_TO_DECIMAL_CAST = castFunctionToDecimalFrom(FLOAT.getTypeSignature(), "floatToShortDecimal", "floatToLongDecimal");
+    public static final SqlScalarFunction DECIMAL_TO_VARCHAR_CAST = castFunctionFromDecimalTo(createUnboundedVarcharType().getTypeSignature(), "shortDecimalToVarchar", "longDecimalToVarchar");
```
Shouldn't this be `varchar(x)`?

haozhun replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java:
```diff
@@ -71,32 +72,32 @@
+    public static final SqlScalarFunction DECIMAL_TO_FLOAT_CAST = castFunctionFromDecimalTo(FLOAT.getTypeSignature(), "shortDecimalToFloat", "longDecimalToFloat");
+    public static final SqlScalarFunction FLOAT_TO_DECIMAL_CAST = castFunctionToDecimalFrom(FLOAT.getTypeSignature(), "floatToShortDecimal", "floatToLongDecimal");
+    public static final SqlScalarFunction DECIMAL_TO_VARCHAR_CAST = castFunctionFromDecimalTo(createUnboundedVarcharType().getTypeSignature(), "shortDecimalToVarchar", "longDecimalToVarchar");
+    public static final SqlScalarFunction VARCHAR_TO_DECIMAL_CAST = castFunctionToDecimalFrom(parseTypeSignature("varchar(y)", ImmutableSet.of("y")), "varcharToShortDecimal", "varcharToLongDecimal");
```
Use `x` instead of `y` for consistency?

haozhun replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/BoundVariables.java:
```diff
@@ -71,9 +71,7 @@ public boolean containsLongVariable(String variableName)
     {
         checkState(variableName != null, "variableName is null");
-        T value = map.get(variableName);
-        checkState(value != null, "value for variable '%s' is null", variableName);
```
Is removing this the right thing to do, or should we be calling contains? Seems like this will remove a layer of validation for functions that implement `specialize`

sopel39 replied:
maybe use `Optional` instead to be more explicit

haozhun replied:
done. I put the validation logic back.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/BindableAggregationFunction.java:
```diff
@@ -93,7 +92,7 @@ public String getDescription()
     public InternalAggregationFunction specialize(BoundVariables variables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
     {
         // bind variables
-        Signature boundSignature = SignatureBinder.applyBoundVariables(getSignature(), variables, arity);
```
Seems like this should have been in the other commit

haozhun replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -445,4 +399,229 @@ private static TypeSignatureParameter applyBoundVariables(TypeSignatureParameter
+        if (variableArityArgumentsCount == 0) {
+            return formalTypeSignatures;
+        }
+        checkArgument(variableArityArgumentsCount > 0 && !formalTypeSignatures.isEmpty());
```
Is this check appropriate? It looks like it's going to cause an `IllegalArgumentException` in `appendArgumentConstraintsTo` if you try to do something like `my_vararg(x, y, z...)` and call it like `my_vararg(1)`

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -445,4 +399,229 @@ private static TypeSignatureParameter applyBoundVariables(TypeSignatureParameter
+                    originalTypeTypeParametersBuilder.add(typeSignatureParameter);
+                }
+            }
+            Type originalType = typeManager.getType(new TypeSignature(formalTypeSignature.getBase(), originalTypeTypeParametersBuilder.build()));
```
I think "current" might be a better name than "original"

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -445,4 +399,229 @@ private static TypeSignatureParameter applyBoundVariables(TypeSignatureParameter
+                }
+                else {
+                    verify(typeSignatureParameter.getKind() == ParameterKind.LONG);
+                    if (commonSuperLongLiteral != typeSignatureParameter.getLongLiteral()) {
```
This relies on the assumption that all literal parameters are upper bounds. That is, for all T with literal parameters, and x < y, `T(x)` is a subtype of `T(y)`. Probably worth putting in a comment mentioning that

martint replied:
That's not an assumption we want to make. It should be based on what implicit coercions are available. In particular, this is not true for decimal types. E.g., `decimal(4,1)` is not implicitly coercible to `decimal(4,2)`

haozhun replied:
I deliberately avoided making such an assumption when writing the initial version. And I don't this assumption in the current version. @cberner, can you elaborate?

cberner replied:
Oh you're right, I didn't think this through carefully enough

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -58,247 +84,270 @@ public SignatureBinder(TypeManager typeManager, Signature declaredSignature, boo
             }
-            default:
-                throw new IllegalStateException("Unknown parameter kind: " + expected.getKind());
+            if (changed) {
```
This check is backwards. It should be `!changed`

haozhun replied:
It's not backwards.

haozhun replied:
done

cberner commented:
Fixes: https://github.com/prestodb/presto/issues/5732

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/BindableAggregationFunction.java:
```diff
@@ -92,7 +93,7 @@ public String getDescription()
     {
         // bind variables
-        Signature boundSignature = bindVariables(getSignature(), variables, arity);
+        Signature boundSignature = SignatureBinder.applyBoundVariables(getSignature(), variables, arity);
```
static import

haozhun replied:
test

haozhun replied:
done

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/FloatSumAggregation.java:
```diff
@@ -20,17 +20,18 @@
 public final class FloatSumAggregation
 {
-    public static final InternalAggregationFunction FLOAT_SUM = generateInternalAggregationFunction(FloatSumAggregation.class);
+    public static final InternalAggregationFunction FLOAT_SUM = generateInternalAggregationFunction(FloatSumAggregation.class, FLOAT.getTypeSignature(), ImmutableList.of(FLOAT.getTypeSignature()));
```
should production code use from `com.facebook.presto.testing.AggregationTestUtils`. It seems that it is test specific.

haozhun replied:
done

sopel39 commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -15,40 +15,66 @@
 
     public SignatureBinder(TypeManager typeManager, Signature declaredSignature, boolean allowCoercion)
     {
+        Map<String, TypeSignature> existingUsages = new HashMap<>();
```
move `existingUsage` variable declaration to `checkNoLiteralVariableUsageAcrossTypes`

sopel39 commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -445,4 +399,229 @@ private static TypeSignatureParameter applyBoundVariables(TypeSignatureParameter
+
+    private static List<TypeSignature> expandVarargFormalTypeSignature(List<TypeSignature> formalTypeSignatures, int actualArity)
+    {
+        int variableArityArgumentsCount = actualArity - formalTypeSignatures.size();
```
shouldn't there be `+1` added?

haozhun replied:
done. You are right. Doing +1 is more consistent with the variable name.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -445,4 +399,229 @@ private static TypeSignatureParameter applyBoundVariables(TypeSignatureParameter
+        }
+    }
+
+    private class TypeRelationshipConstraintSolver
```
Is it really a solver? It is more like an assertion. If it doesn't update bindings, then how it can be satisfied? If other solvers update bindings so that `TypeRelationshipConstraintSolver` is satisfied, then why they can't have conditions from `TypeRelationshipConstraintSolver` in themselves/

sopel39 commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -445,4 +399,229 @@ private static TypeSignatureParameter applyBoundVariables(TypeSignatureParameter
+        @Override
+        public boolean isSatisfied(BoundVariables.Builder bindings)
+        {
+            return true;
```
shouldn't you at least assert that the bindings are really correct?

haozhun replied:
Removed this method

sopel39 commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -445,4 +399,229 @@ private static TypeSignatureParameter applyBoundVariables(TypeSignatureParameter
+                    Long typeParameter = bindings.getLongVariable(typeSignatureParameter.getVariable());
+                    if (typeParameter == null) {
+                        // if an existing value doesn't exist for the given variable name, use the value that comes from the actual type.
+                        Optional<Type> type = typeManager.coerceTypeBase(actualType, formalTypeSignature.getBase());
```
I think long parameters from `coerceTypeBase` should be used exclusively (e.g not mixed with existing bindings or provided long parameters). For example:
`function(decimal(x, 10))`
and you call it: `function(BIGINT 10)`. Then `x` would be bound `19` and the function signature would be `function(decimal(19,10))`. This is incorrect.

Maybe make such arguments like: `decimal(x,10)` illegal?

haozhun replied:
This example you gave works. It will give you `decimal(29,10)` correctly.

But there are cases where the code can't handle. I added them to javadoc of this class.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -445,4 +399,229 @@ private static TypeSignatureParameter applyBoundVariables(TypeSignatureParameter
+        @Override
+        public boolean isSatisfied(BoundVariables.Builder bindings)
+        {
+            return true;
```
assert?

haozhun replied:
Removed this method

sopel39 commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -63,242 +89,265 @@ public SignatureBinder(TypeManager typeManager, Signature declaredSignature, boo
+
+        calculateVariableValuesForLongConstraints(boundVariablesBuilder);
+
+        BoundVariables boundVariables = boundVariablesBuilder.build();
```
I wonder if variables builder should accept updates. Guava builders don't accept duplicate keys. Maybe plain map should be used for solving variables

haozhun replied:
Plain map does not work very well because of the distinction between type variables and long variables. If having a builder that allows overwriting is surprising to you, maybe we can discuss about changing its name. I would also like to point out that I didn't change the behavior of BoundVariables (now that I undid my change to its getters).

sopel39 commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -15,40 +15,66 @@
+ */
 public class SignatureBinder
 {
+    private static final int SOLVE_ITERATION_LIMIT = 4;
```
why 4? Can you give an example where iterations > 1 are actually needed?

haozhun replied:
With the iterative solver, you need 2 for anything with generic in it. For expressions involving lambda (which is under progress), you will need 3.

The number here is chosen rather arbitrarily. The purpose of this constant is to avoid infinite loop. I'll add comments.

sopel39 replied:
please do. An example would be great.

haozhun commented:
I addressed the comments.

Here are  the larger changes I made and my justifications:
* remove `isSatisfied` method
  * This method is confusing in that the method assumes `update` had previously returned `UNCHANGED`.
  * I changed update to return `UNCHANGED_SATISFIED`, `UNCHANGED_NOT_SATISFIED`, `CHANGED`, `UNSOLVABLE`.
* rename `UNCHANGED` to `UNCHANGED_SATISFIED`
  * In our draft, we had bugs where we return `UNCHANGED` when we should return `UNSOLVABLE`. I find this partially caused by the improper name of `UNCHANGED`.
* change `TypeRelationshipConstraintSolver` to be 3-ways instead of 2-ways

haozhun commented:
The Travis failure seems unrelated.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -325,133 +418,291 @@ private void calculateVariableValuesForLongConstraints(BoundVariables.Builder va
     }
 
-    private static List<TypeSignature> fillInMissingVariableArguments(List<TypeSignature> argumentSignatures, int arity)
+    private class SolverReturnStatusMerger
```
why do you need a separate class for this logic? Wouldn't simple method `mergeSolverReturnStatuses` be enough?

haozhun replied:
good point

sopel39 commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -325,133 +418,291 @@ private void calculateVariableValuesForLongConstraints(BoundVariables.Builder va
+                return SolverReturnStatus.CHANGED;
+            }
+            Type originalType = bindings.getTypeVariable(typeParameter);
+            Optional<Type> commonSuperType = typeManager.getCommonSuperType(originalType, actualType);
```
As I pointed yesterday to @cberner and @martint, it might be needed for `getCommonSuperType` to behave differently in SQL context and lambda context. In SQL context character and possibly numeric types probably shouldn't have supertype (e. g. `char(5)` is not a supertype of `char(3)`, `double` is not a supertype of `decimal`). The difference in behavior is driven by SQL 99 standard compatibility.
